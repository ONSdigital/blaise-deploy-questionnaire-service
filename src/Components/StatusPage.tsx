import React, {Fragment, ReactElement, useEffect, useState} from "react";
import {Link} from "react-router-dom";
import {ErrorBoundary} from "./ErrorHandling/ErrorBoundary";


function StatusPage(): ReactElement {
    const [statusList, setStatusList] = useState<any[]>([]);
    const [listError, setListError] = useState<string>("Loading ...");

    useEffect(() => {
        getHeathStatus();
    }, []);

    function getHeathStatus() {
        setStatusList([]);
        console.log("getHeathStatus");
        fetch("/api/health")
            .then((r: Response) => {
                if (r.status !== 503 && r.status !== 200) {
                    throw r.status + " - " + r.statusText;
                }
                r.json()
                    .then((json: any) => {
                        if (!Array.isArray(json)) {
                            throw "Json response is not a list";
                        }
                        console.log("Retrieved blaise status list, " + json.length + " items/s");
                        console.log(json);
                        setStatusList(json);
                        setListError("");
                        if (json.length === 0) setListError("No connection details found.");
                    })
                    .catch((error) => {
                        console.error("Unable to read json from response, error: " + error);
                        setListError("Unable to get Blaise status");
                    });
            }).catch((error) => {
                console.error("Failed to retrieve Blaise status, error: " + error);
                setListError("Unable to get Blaise status");
            }
        );
    }


    return (
        <>
            <p>
                <Link to={"/"}>Previous</Link>
            </p>
            <h1>Blaise connection status</h1>
            <ErrorBoundary errorMessageText={"Failed to load Blaise Status, best to assume its not working."}>
                {
                    listError !== "" ?
                        <p>{listError}</p>
                        :
                        <dl className="metadata metadata__list grid grid--gutterless u-cf u-mb-l"
                            title="Status information for connections to Blaise"
                            aria-label="Status information for connections to Blaise">
                            {statusList.map((item: any) => {
                                return (
                                    <Fragment key={item["health check type"]}>
                                        <dt className="metadata__term grid__col col-5@m">
                                            {item["health check type"]}:
                                        </dt>
                                        <dd className="metadata__value grid__col col-7@m">
                                             <span
                                                 className={`status status--${(item.status === "OK" ? "success" : "error")}`}>
                                                 {item.status}
                                             </span>
                                        </dd>
                                    </Fragment>
                                );
                            })}
                        </dl>
                }
            </ErrorBoundary>
        </>
    );
}

export default StatusPage;
