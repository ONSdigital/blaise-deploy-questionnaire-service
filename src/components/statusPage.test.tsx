/**
 * @jest-environment jsdom
 */

import React from "react";
import { render, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import flushPromises from "../tests/utils";
import { act } from "react-dom/test-utils";
import StatusPage from "./statusPage";
import { MemoryRouter } from "react-router-dom";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";

const mock = new MockAdapter(axios);

const status_list = [
    {
        "health check type": "Connection model",
        "status": "OK"
    },
    {
        "health check type": "Blaise connection",
        "status": "OK"
    },
    {
        "health check type": "Remote data server connection",
        "status": "OK"
    },
    {
        "health check type": "Remote Cati management connection",
        "status": "OK"
    }
];

describe("Blaise Status page", () => {
    beforeEach(() => {
        mock.onGet("/api/health/diagnosis").reply(200, status_list);
    });

    afterEach(() => {
        mock.reset();
    });

    it("view Blaise Status page matches Snapshot", async () => {

        const wrapper = render(
            <MemoryRouter>
                <StatusPage />
            </MemoryRouter>
        );

        await act(async () => {
            await flushPromises();
        });

        await waitFor(() => {
            expect(wrapper).toMatchSnapshot();
        });
    });

    it("should render correctly", async () => {

        const { getByText, queryByText } = render(
            <MemoryRouter>
                <StatusPage />
            </MemoryRouter>
        );

        expect(queryByText(/Checking Blaise status/i)).toBeInTheDocument();

        await waitFor(() => {
            expect(getByText(/Blaise connection status/i)).toBeDefined();
            expect(queryByText(/Checking Blaise status/i)).not.toBeInTheDocument();
        });

        await waitFor(() => {
            expect(getByText(/Blaise connection status/i)).toBeDefined();
            expect(getByText(/Connection model/i)).toBeDefined();
            expect(getByText(/Remote Cati management connection/i)).toBeDefined();
            expect(queryByText(/Checking Blaise status/i)).not.toBeInTheDocument();
        });

    });
});

describe("Given the API returns a 500 status", () => {
    beforeEach(() => {
        mock.onGet("/api/health/diagnosis").reply(500);
    });

    afterEach(() => {
        mock.reset();
    });

    it("it should render with the error message displayed", async () => {

        const { getByText, queryByText } = render(
            <MemoryRouter>
                <StatusPage />
            </MemoryRouter>
        );

        expect(queryByText(/Checking Blaise status/i)).toBeInTheDocument();

        await waitFor(() => {
            expect(getByText(/Unable to get Blaise status/i)).toBeDefined();
            expect(queryByText(/Checking Blaise status/i)).not.toBeInTheDocument();
        });
    });
});

describe("Given the API returns malformed json", () => {
    beforeEach(() => {
        mock.onGet("/api/health/diagnosis").reply(200, { text: "Hello" });
    });

    afterEach(() => {
        mock.reset();
    });

    it("it should render with the error message displayed", async () => {

        const { getByText, queryByText } = render(
            <MemoryRouter>
                <StatusPage />
            </MemoryRouter>
        );

        expect(queryByText(/Checking Blaise status/i)).toBeInTheDocument();

        await waitFor(() => {
            expect(getByText(/Unable to get Blaise status/i)).toBeDefined();
            expect(queryByText(/Checking Blaise status/i)).not.toBeInTheDocument();
        });
    });
});

describe("Given the API returns an empty list", () => {
    beforeEach(() => {
        mock.onGet("/api/health/diagnosis").reply(200, []);
    });

    afterEach(() => {
        mock.reset();
    });

    it("it should render with a message to inform the user in the list", async () => {

        const { getByText, queryByText } = render(
            <MemoryRouter>
                <StatusPage />
            </MemoryRouter>
        );

        expect(queryByText(/Checking Blaise status/i)).toBeInTheDocument();

        await waitFor(() => {
            expect(getByText(/No connection details found./i)).toBeDefined();
            expect(queryByText(/Checking Blaise status/i)).not.toBeInTheDocument();
        });
    });
});
