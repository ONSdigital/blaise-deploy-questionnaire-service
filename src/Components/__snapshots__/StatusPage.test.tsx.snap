// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blaise Status page view Blaise Status page matches Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <p>
        <a
          href="/"
        >
          Previous
        </a>
      </p>
      <h1>
        Blaise connection status
      </h1>
      <dl
        aria-label="Status information for connections to Blaise"
        class="metadata metadata__list grid grid--gutterless u-cf u-mb-l"
        title="Status information for connections to Blaise"
      >
        <dt
          class="metadata__term grid__col col-5@m"
        >
          Connection model
          :
        </dt>
        <dd
          class="metadata__value grid__col col-7@m"
        >
          <span
            class="status status--success"
          >
            OK
          </span>
        </dd>
        <dt
          class="metadata__term grid__col col-5@m"
        >
          Blaise connection
          :
        </dt>
        <dd
          class="metadata__value grid__col col-7@m"
        >
          <span
            class="status status--success"
          >
            OK
          </span>
        </dd>
        <dt
          class="metadata__term grid__col col-5@m"
        >
          Remote data server connection
          :
        </dt>
        <dd
          class="metadata__value grid__col col-7@m"
        >
          <span
            class="status status--success"
          >
            OK
          </span>
        </dd>
        <dt
          class="metadata__term grid__col col-5@m"
        >
          Remote Cati management connection
          :
        </dt>
        <dd
          class="metadata__value grid__col col-7@m"
        >
          <span
            class="status status--success"
          >
            OK
          </span>
        </dd>
      </dl>
    </div>
  </body>,
  "container": <div>
    <p>
      <a
        href="/"
      >
        Previous
      </a>
    </p>
    <h1>
      Blaise connection status
    </h1>
    <dl
      aria-label="Status information for connections to Blaise"
      class="metadata metadata__list grid grid--gutterless u-cf u-mb-l"
      title="Status information for connections to Blaise"
    >
      <dt
        class="metadata__term grid__col col-5@m"
      >
        Connection model
        :
      </dt>
      <dd
        class="metadata__value grid__col col-7@m"
      >
        <span
          class="status status--success"
        >
          OK
        </span>
      </dd>
      <dt
        class="metadata__term grid__col col-5@m"
      >
        Blaise connection
        :
      </dt>
      <dd
        class="metadata__value grid__col col-7@m"
      >
        <span
          class="status status--success"
        >
          OK
        </span>
      </dd>
      <dt
        class="metadata__term grid__col col-5@m"
      >
        Remote data server connection
        :
      </dt>
      <dd
        class="metadata__value grid__col col-7@m"
      >
        <span
          class="status status--success"
        >
          OK
        </span>
      </dd>
      <dt
        class="metadata__term grid__col col-5@m"
      >
        Remote Cati management connection
        :
      </dt>
      <dd
        class="metadata__value grid__col col-7@m"
      >
        <span
          class="status status--success"
        >
          OK
        </span>
      </dd>
    </dl>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`React homepage view Blaise Status page matches Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <p>
        <a
          href="/"
        >
          Previous
        </a>
      </p>
      <h1>
        Blaise connection status
      </h1>
      <dl
        aria-label="Status information for connections to Blaise"
        class="metadata metadata__list grid grid--gutterless u-cf u-mb-l"
        title="Status information for connections to Blaise"
      >
        <dt
          class="metadata__term grid__col col-5@m"
        >
          Connection model
          :
        </dt>
        <dd
          class="metadata__value grid__col col-7@m"
        >
          <span
            class="status status--success"
          >
            OK
          </span>
        </dd>
        <dt
          class="metadata__term grid__col col-5@m"
        >
          Blaise connection
          :
        </dt>
        <dd
          class="metadata__value grid__col col-7@m"
        >
          <span
            class="status status--success"
          >
            OK
          </span>
        </dd>
        <dt
          class="metadata__term grid__col col-5@m"
        >
          Remote data server connection
          :
        </dt>
        <dd
          class="metadata__value grid__col col-7@m"
        >
          <span
            class="status status--success"
          >
            OK
          </span>
        </dd>
        <dt
          class="metadata__term grid__col col-5@m"
        >
          Remote Cati management connection
          :
        </dt>
        <dd
          class="metadata__value grid__col col-7@m"
        >
          <span
            class="status status--success"
          >
            OK
          </span>
        </dd>
      </dl>
    </div>
  </body>,
  "container": <div>
    <p>
      <a
        href="/"
      >
        Previous
      </a>
    </p>
    <h1>
      Blaise connection status
    </h1>
    <dl
      aria-label="Status information for connections to Blaise"
      class="metadata metadata__list grid grid--gutterless u-cf u-mb-l"
      title="Status information for connections to Blaise"
    >
      <dt
        class="metadata__term grid__col col-5@m"
      >
        Connection model
        :
      </dt>
      <dd
        class="metadata__value grid__col col-7@m"
      >
        <span
          class="status status--success"
        >
          OK
        </span>
      </dd>
      <dt
        class="metadata__term grid__col col-5@m"
      >
        Blaise connection
        :
      </dt>
      <dd
        class="metadata__value grid__col col-7@m"
      >
        <span
          class="status status--success"
        >
          OK
        </span>
      </dd>
      <dt
        class="metadata__term grid__col col-5@m"
      >
        Remote data server connection
        :
      </dt>
      <dd
        class="metadata__value grid__col col-7@m"
      >
        <span
          class="status status--success"
        >
          OK
        </span>
      </dd>
      <dt
        class="metadata__term grid__col col-5@m"
      >
        Remote Cati management connection
        :
      </dt>
      <dd
        class="metadata__value grid__col col-7@m"
      >
        <span
          class="status status--success"
        >
          OK
        </span>
      </dd>
    </dl>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
